packages/tpetra/core/src/Tpetra_Import_def.hpp

```diff
@@ -1064,205 +1077,224 @@ namespace Tpetra {
+
+	  // Ignore remote GIDs that aren't owned by any process in the
+	  // source Map.  getRemoteIndexList() gives each of these a
+	  // process ID of -1.
+
+	  const size_type numInvalidRemote =
+	      std::count_if (remoteProcIDs.begin (), remoteProcIDs.end (),
+			     std::bind1st (std::equal_to<int> (), -1));
+	  TEUCHOS_TEST_FOR_EXCEPTION_CLASS_FUNC
+	      (numInvalidRemote == 0, std::logic_error, "Calling getRemoteIndexList "
+	       "on the source Map returned IDNotPresent, but none of the returned "
+	       "\"remote\" process ranks are -1.  Please report this bug to the "
+	       "Tpetra developers.");
+
+#ifdef HAVE_TPETRA_MMM_TIMINGS
+	  using Teuchos::TimeMonitor;
+	  std::string label;
+	  if(!plist.is_null())
+	      label = plist->get("Timer Label",label);
+	  std::string prefix = std::string("Tpetra ")+ label + std::string(":iport_ctor:setupExport:1 ");
+	  auto MM = rcp(new TimeMonitor(*TimeMonitor::getNewTimer(prefix)));
+#endif
+
+
+	  // If all of them are invalid, we can delete the whole array.
+	  const size_type totalNumRemote = getNumRemoteIDs ();
+	  if (numInvalidRemote == totalNumRemote) {
+	      // all remotes are invalid; we have no remotes; we can delete the remotes
+	      remoteProcIDs.clear ();
+	      remoteGIDs.clear (); // This invalidates the view remoteGIDsView
+	      ImportData_->remoteLIDs_.clear();
+	  }
+	  else {
+	      // Some remotes are valid; we need to keep the valid ones.
+	      // Pack and resize remoteProcIDs, remoteGIDs, and remoteLIDs_.
+	      size_type numValidRemote = 0;
 #ifdef HAVE_TPETRA_DEBUG
-        ArrayView<GlobalOrdinal> remoteGIDsPtr = remoteGIDsView;
+	      ArrayView<GlobalOrdinal> remoteGIDsPtr = remoteGIDsView;
 #else
-        GlobalOrdinal* const remoteGIDsPtr = remoteGIDsView.getRawPtr ();
+	      GlobalOrdinal* const remoteGIDsPtr = remoteGIDsView.getRawPtr ();
 #endif // HAVE_TPETRA_DEBUG
-        for (size_type r = 0; r < totalNumRemote; ++r) {
-          // Pack in all the valid remote PIDs and GIDs.
-          if (remoteProcIDs[r] != -1) {
-            remoteProcIDs[numValidRemote] = remoteProcIDs[r];
-            remoteGIDsPtr[numValidRemote] = remoteGIDsPtr[r];
-            ImportData_->remoteLIDs_[numValidRemote] = ImportData_->remoteLIDs_[r];
-            ++numValidRemote;
-          }
-        }
-        TEUCHOS_TEST_FOR_EXCEPTION(
-          numValidRemote != totalNumRemote - numInvalidRemote, std::logic_error,
-          "Tpetra::Import::setupExport(): After removing invalid remote GIDs and"
-          " packing the valid remote GIDs, numValidRemote = " << numValidRemote
-          << " != totalNumRemote - numInvalidRemote = "
-          << totalNumRemote - numInvalidRemote
-          << ".  Please report this bug to the Tpetra developers.");
-
-        remoteProcIDs.resize (numValidRemote);
-        remoteGIDs.resize (numValidRemote);
-        ImportData_->remoteLIDs_.resize (numValidRemote);
+	      for (size_type r = 0; r < totalNumRemote; ++r) {
```

Comment:
It looks like not much has changed in the code below, other than formatting and timers.  Is that right?